### PR TITLE
Treatment enum

### PR DESCRIPTION
**Pull request by mBialczak**
Introduced Treatment enum class for use in Visit.

Some fix in visit tests was necessary - test behaved randomly, which did not reveal itself until completely unrelated tests for enum where added.

### CODE REVIEW

File: src/app/visit/visit.cpp [hunk 1]
```diff
@@ -57,3 +63,8 @@ std::string Visit::getVisitInformation() const
 {
     return visit_information_;
 }
+
+void Visit::updateTreatments(std::vector<Treatment> newTreatments)
+{
+    treatments_ = std::move(newTreatments);
+}
```
Comment (fbtom) [inline src/app/visit/visit.cpp:70]: EOF add enter

File: src/app/visit/visit.hpp [hunk 2]
```diff
@@ -1,25 +1,33 @@
 #pragma once
 
+#include "treatment.hpp"
+
```
Comment (fbtom) [inline src/app/visit/visit.hpp:4]: remove enter

File: src/tests/ut/visit/treatment_test.cpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+TEST_P(TreatmentParameterizedFixture, GivenTreatmentEnumToStringShouldProvideCorrectStringRepresentation)
+{
+
+    const auto& [enumValue, stringRepresentation] = GetParam();
+
+    EXPECT_EQ(toString(enumValue), stringRepresentation);
+}
+
+INSTANTIATE_TEST_SUITE_P(TreatmentConversionToStringTest, TreatmentParameterizedFixture,
+                         ::testing::ValuesIn(expectedDentalTreatments));
```
Comment (fbtom) [inline src/tests/ut/visit/treatment_test.cpp:25]: This way it's hidden what is being tested and it's harder to read and understand 
Please follow apprach with TestParam 
and as follows `TestParams{Treatments::Dentures, "dentures"}`

File: src/app/visit/treatment.cpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
+#include "treatment.hpp"
+
+#include <map>
+#include <stdexcept>
+
+namespace
+{
+const std::map<Treatment, std::string> DentalTreatments = {
+    {Treatment::TeethCleaning, "Teeth Cleaning"},   {Treatment::RootCanal, "Root Canal"},
```
Comment (fbtom) [inline src/app/visit/treatment.cpp:9]: In my opition we should provide lowercase strings and later on provide layer that will take care of formatting when needed

File: src/tests/ut/visit/visit_test.cpp [hunk 5]
```diff
@@ -2,47 +2,63 @@
...
-        else
-        {
-            content = "null";
-        }
-    }
+
     auto expected = "I w klubie sa sami fajni ludzie";
-    EXPECT_EQ(content, expected);
+
+    EXPECT_TRUE(std::ranges::any_of(
```
Comment (fbtom) [inline src/tests/ut/visit/visit_test.cpp:37]: Within gtest macros I prefer to see only call such as "getXvalue" and compare, please avoid running algorithms inside. 
You can introduce `const auto any_visit_information = std::ranges...`
And then `EXPECT_TRUE(any_visit_information)`
Proposed naming might be improved

File: src/tests/ut/visit/visit_test.cpp [hunk 6]
```diff
@@ -2,47 +2,63 @@
...
     auto content = doc2->getVisitAssociations().begin()->get()->getVisitInformation();
     auto expected = "Tworze klub ninja!";
     EXPECT_EQ(content, expected);
 }
 
-TEST_F(AssociationTest, AssociationDemo3)
+TEST_F(VisitTestsFixture, AssociationDemo3)
 {
     std::string content;
     auto associations = doc2->getVisitAssociations();
```
Comment (fbtom) [inline src/tests/ut/visit/visit_test.cpp:33]: As you already change something within those tests, please consider switching auto to const auto where possible.

File: src/tests/ut/visit/visit_test.cpp [hunk 7]
```diff
@@ -2,47 +2,63 @@
...
     auto expected = "I w klubie sa sami fajni ludzie";
-    EXPECT_EQ(content, expected);
+
+    EXPECT_TRUE(std::ranges::any_of(
+        associations, [&expected](const auto& visitSPtr) { return visitSPtr->getVisitInformation() == expected; }));
+}
+
+TEST_F(VisitTestsFixture, UpdateTreatmentsShouldReplacePreviouslyStoredTreatments)
+{
+    auto visitOneTreatmentsBefore = visit1->getTreatments();
```
Comment (fbtom) [inline src/tests/ut/visit/visit_test.cpp:43]: const auto

File: src/tests/ut/visit/visit_test.cpp [hunk 8]
```diff
@@ -2,47 +2,63 @@
...
+    auto visitTwoTreatmentsBefore = visit2->getTreatments();
+    auto visitThreeTreatmentsBefore = visit3->getTreatments();
+    // initial asserts
```
Comment (fbtom) [inline src/tests/ut/visit/visit_test.cpp:46]: Comment is not needed here, make test name more descriptive if this has to be marked

File: src/tests/ut/visit/visit_test.cpp [hunk 9]
```diff
@@ -2,47 +2,63 @@
...
+    ASSERT_TRUE(visitOneTreatmentsBefore.empty());
+    ASSERT_EQ(visitTwoTreatmentsBefore, std::vector{Treatment::TeethWhitening});
+    ASSERT_EQ(visitThreeTreatmentsBefore,
+              (std::vector{Treatment::TeethCleaning, Treatment::RootCanal, Treatment::ToothExtraction}));
+
+    // update treatments for visit
```
Comment (fbtom) [inline src/tests/ut/visit/visit_test.cpp:52]: same here
and consider creating a method in fixture to do such updates

File: src/tests/ut/visit/visit_test.hpp [hunk 10]
```diff
@@ -5,7 +5,7 @@
 
 class Doctor;
 
-struct AssociationTest : ::testing::Test
+struct VisitTestsFixture : ::testing::Test
```
Comment (fbtom) [inline src/tests/ut/visit/visit_test.hpp:8]: please switch to class and use `protected`

**Review by fbtom — CHANGES_REQUESTED** [review]

**Comment by ziobron** [comment]
Thanks @fbtom for Code Review.
🏅 1 XP granted

File: src/app/visit/treatment.cpp [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    {Treatment::DentalFilling, "Dental Filling"},   {Treatment::ToothExtraction, "Tooth Extraction"},
+    {Treatment::TeethWhitening, "Teeth Whitening"}, {Treatment::DentalImplants, "Dental Implants"},
+    {Treatment::Orthodontics, "Orthodontics"},      {Treatment::Periodontics, "Periodontics"},
+    {Treatment::DentalVeneers, "Dental Veneers"},   {Treatment::DentalCrowns, "Dental Crowns"},
+    {Treatment::DentalBridges, "Dental Bridges"},   {Treatment::Dentures, "Dentures"},
+    {Treatment::GumSurgery, "Gum Surgery"},         {Treatment::DentalSealants, "Dental Sealants"},
+    {Treatment::MouthGuards, "Mouth Guards"},       {Treatment::SleepApneaTreatment, "Sleep Apnea Treatment"}};
+} // namespace
```
Comment (BaqWin) [inline src/app/visit/treatment.cpp:17]: Maybe think about adding a default treatment or call it "Other" if the treatment isn't in our enum

**Comment by ziobron** [comment]
Thanks @BaqWin for Code Review.
🏅 1 XP granted

File: src/app/visit/treatment.hpp [hunk 12]
```diff
@@ -0,0 +1,25 @@
+#pragma once
+
+#include <string>
+
+enum class Treatment : unsigned
+{
+    TeethCleaning,
```
[the remark this answers is not in the document]
Comment (mBialczak) [reply src/app/visit/treatment.hpp]: done

File: src/app/visit/visit.cpp (hunk 1 above)
Comment (mBialczak) [reply src/app/visit/visit.cpp]: done

File: src/app/visit/visit.hpp (hunk 2 above)
Comment (mBialczak) [reply src/app/visit/visit.hpp]: done

File: src/tests/ut/visit/treatment_test.cpp (hunk 3 above)
Comment (mBialczak) [reply src/tests/ut/visit/treatment_test.cpp]: Done with name TestParams (but I left creating values to one place in container}

File: src/tests/ut/visit/visit_test.cpp (hunk 6 above)
Comment (mBialczak) [reply src/tests/ut/visit/visit_test.cpp]: Replaced auto with const auto& wherever possible (or not missed :D)

File: src/tests/ut/visit/visit_test.cpp (hunk 7 above)
Comment (mBialczak) [reply src/tests/ut/visit/visit_test.cpp]: done

File: src/tests/ut/visit/visit_test.cpp (hunk 8 above)
Comment (mBialczak) [reply src/tests/ut/visit/visit_test.cpp]: Comments removed. 
Single update function implemented in test fixture.

File: src/tests/ut/visit/visit_test.cpp (hunk 9 above)
Comment (mBialczak) [reply src/tests/ut/visit/visit_test.cpp]: Single update function implemented in test fixture.
Comment removed.

File: src/app/visit/treatment.cpp (hunk 11 above)
Comment (mBialczak) [reply src/app/visit/treatment.cpp]: Added "OTHER" value to enum.

File: src/app/visit/visit.cpp [hunk 13]
```diff
@@ -13,10 +13,11 @@ Visit::~Visit()
     removeFromExtent(this);
 }
 
-std::shared_ptr<Visit> Visit::createVisit(std::shared_ptr<Doctor> doctor)
+std::shared_ptr<Visit> Visit::createVisit(std::shared_ptr<Doctor> doctor, std::vector<Treatment> predictedTreatments)
```
[the remark this answers is not in the document]
Comment (mBialczak) [reply src/app/visit/visit.cpp]: done

File: src/tests/ut/visit/visit_test.cpp (hunk 6 above)
Comment (fbtom) [reply src/tests/ut/visit/visit_test.cpp]: For "things" that are created on-fly such as:
`const auto& expected = 3;`
we shouldn't use & as there might be an issue with gtest macro resolving (if I remember correctly) 
Would be good to adjust that to `const auto` and the other places, where we just copy values would be fine too imo

**Comment by ziobron** [comment]
Thanks @fbtom for Code Review.
🏅 1 XP granted

File: src/tests/ut/visit/visit_test.hpp (hunk 10 above)
Comment (fbtom) [reply src/tests/ut/visit/visit_test.hpp]: Ah, I meant a protected keyword inside as you did. Inheritance has to be public, well done :)

**Comment by ziobron** [comment]
Thanks @fbtom for Code Review.
🏅 1 XP granted

File: src/tests/ut/visit/visit_test.cpp (hunk 6 above)
Comment (mBialczak) [reply src/tests/ut/visit/visit_test.cpp]: You are completely right. I got ahead of myself and didn't even check those all are copies. Thanks for spotting that!

**Review by fbtom — APPROVED** [review]

**Comment by ziobron** [comment]
Thanks @fbtom for Code Review.
🏅 1 XP granted

**Comment by ziobron** [comment]
Your PR was merged!
🏅 2 XP granted. Thanks @mBialczak for making course materials better!

**Pull request merged**